configfile: 'config/config.yaml'

include: 'rules/genotyping.smk'
include: 'rules/discovery.smk'
include: 'rules/common-rules.smk'
include: 'rules/get-sample-list.smk'

if config['pilot']:
    samples=['HG00099']

# remove problematic samples from the snakemake pipeline
#bad_samples = ['HG01699', 'HG00139', 'HG00235', 'HG00268', 'HG01915', 'HG01791', 'HG01879', 'HG03009', 'HG03521']
#samples = [s for s in samples if s not in bad_samples]

wildcard_constraints:
    sample='|'.join(samples)

rule genotype_experiment:
    input:
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/svarp-giggles/chm13-90c.r518/genotypes/{sample}-biallelic.vcf.gz', sample=samples),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/svarp-giggles/chm13-90c.r518/genotypes/{sample}-multiallelic.vcf.gz', sample=samples),
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/svarp-giggles/chm13-90c.r518/genotypes/multisample-multiallelic.vcf.gz',
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/svarp-giggles/chm13-90c.r518/genotypes/multisample-biallelic.vcf.gz',
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/svarp-giggles/chm13-90c.r518/plots/callset-stats.tsv',
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/svarp-giggles/chm13-90c.r518/plots/sv_count.tsv'
        
rule svarp_experiment:
    input:
        #expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/svarp-giggles/chm13-90c.r518/svarp/{sample}/svimasm/variants.vcf', sample=samples),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/svarp-giggles/chm13-90c.r518/svarp/{sample}/pav_{ref}/pav_svtigs_merged.vcf', ref=['hg38','t2t'], sample=samples)


rule all:
    input:
        rules.genotype_experiment.input,
        rules.svarp_experiment.input
