"""
Haplotagging all the reads in the CRAM files against the Phased NYGC VCFs.
"""
configfile: './config.yaml'

import subprocess
import re
import gzip
import pandas as pd

#Finding sample list used for 1000GP project by us
path='/gpfs/project/projects/medbioinf/data/share/globus/1000g-ont/GRCh38/'
ls_command = 'ls '+path
process = subprocess.Popen(ls_command.split(), stdout=subprocess.PIPE)
ls_out, ls_err = process.communicate()
sample_re = re.compile('(?:NA|HG)\d{5}')

cram_sample_list = [s for s in ls_out.decode('utf-8').split('\n') if sample_re.match(s)]

vcf_sample_list = None
with gzip.open("/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/nygc-merged.vcf.gz", "rt") as f:
    for line in f:
        if line[0:2] == "##":
            continue
        vcf_sample_list = line.rstrip().split("\t")[9:]
        break

common_samples = list(set(cram_sample_list).intersection(vcf_sample_list))

master_list = "/gpfs/project/projects/medbioinf/users/spani/results/1000GP/1k_ont_data_overview-samples_and_sequencing.tsv"
df = pd.read_csv(master_list, sep="\t")
df = df[["SAMPLE","SEX"]].to_numpy()
female_samples = []
male_samples = []
unspecified_samples = []
for sample,sex in df:
    if sex == "Female":
        female_samples.append(sample)
    elif sex == "Male":
        male_samples.append(sample)
    else:
        unspecified_samples.append(sample) #HG03895 sex is not known.

if config['mode'] == 'pilot':
    all_samples = config['HGSVC'] + config['HPRC']
    female_samples = list(set(all_samples).intersection(female_samples))
    male_samples = list(set(all_samples).intersection(male_samples))
    unspecified_samples = list(set(all_samples).intersection(unspecified_samples))
elif config['mode'] == 'all':
    all_samples = common_samples
    female_samples = list(set(common_samples).intersection(female_samples))
    male_samples = list(set(common_samples).intersection(male_samples))
    unspecified_samples = list(set(common_samples).intersection(unspecified_samples))


wildcard_constraints:
    chr="chr[0-9A-Z]"

#############################
##### Rule Definitions: #####
#############################

rule all:
    input:
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/{sample}/{sample}.tsv', sample=female_samples),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/{sample}/{sample}.tsv', sample=male_samples),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/{sample}/{sample}.tsv', sample=unspecified_samples),
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/sample_sex.tsv'
        

############################
##### Merge NYGC VCFs ######
############################

rule create_region_vcf_file_list:
    input:
        vcfs=expand("/gpfs/project/projects/medbioinf/users/spani/files/vcf/1000GP/NYGC/phased/1kGP_high_coverage_Illumina.{chr}.filtered.SNV_INDEL_SV_phased_panel.vcf.gz", chr=config['chromosome'])
    output:
        temp('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/nygc-vcf-list.txt')
    run:
        f = open(output[0], 'w')
        for name in input.vcfs:
            print(name, file=f)
        f.close()

rule concat_nygc_vcf:
    input:
        file_list='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/nygc-vcf-list.txt'
    output:
        vcf='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/nygc-merged.vcf.gz',
        ind='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/nygc-merged.vcf.gz.csi'
    log:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/nygc-merged.log'
    conda:
        'envs/preprocessing.yaml'
    resources:
        runtime_hrs=5,
        runtime_min=0,
        mem_total_mb=lambda wildcards, attempt: 2000 * attempt
    shell:
        '''
        bcftools concat -o {output.vcf} -Oz -f {input.file_list} &> {log}
        bcftools index {output.vcf}
        '''

########################################
##### Split Phased VCF Sample-wise #####
########################################

rule split_phased_vcf:
    input:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/nygc-merged.vcf.gz'
    output:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}.vcf.gz'
    wildcard_constraints:
        sample = "|".join(all_samples)
    conda:
        'envs/preprocessing.yaml'
    resources:
        runtime_hrs=8,
        runtime_min=0,
        mem_total_mb=lambda wildcards, attempt: 2000 * attempt
    shell:
        '''
        bcftools view -Oz -s {wildcards.sample} -o {output} {input}
        '''

rule index_phased_vcf:
    input:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}.vcf.gz'
    output:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}.vcf.gz.csi'
    wildcard_constraints:
        sample = "|".join(all_samples)
    conda:
        'envs/preprocessing.yaml'
    resources:
        runtime_hrs=0,
        runtime_min=20,
        mem_total_mb=lambda wildcards, attempt: 2000 * attempt
    shell:
        '''
        bcftools index {input}
        '''

rule extract_PAR:
    input:
        vcf='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}.vcf.gz',
        script='/gpfs/project/projects/medbioinf/users/spani/scripts/1000g-ont/ont-1kg/snakemake_pipelines/haplotagging/scripts/extract_PAR.py'
    output:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}_PAR_only.vcf.gz'
    wildcard_constraints:
        sample = "|".join(all_samples)
    conda:
        'envs/basic.yaml'
    resources:
        runtime_hrs=1,
        runtime_min=20,
        mem_total_mb=lambda wildcards, attempt: 2000 * attempt
    shell:
        '''
        python {input.script} --bgzip --vcf {input.vcf} --output {output}
        '''

rule index_PARonly_vcf:
    input:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}_PAR_only.vcf.gz'
    output:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}_PAR_only.vcf.gz.csi'
    wildcard_constraints:
        sample = "|".join(all_samples)
    conda:
        'envs/basic.yaml'
    resources:
        runtime_hrs=0,
        runtime_min=20,
        mem_total_mb=lambda wildcards, attempt: 2000 * attempt
    shell:
        '''
        bcftools index {input}
        '''


##############################
##### Haplotagging Reads #####
##############################

# converting CRAMs to BAMs (since the CRAMs were creating some issues)
rule convert_cram_to_bam:
    input:
        reads='/gpfs/project/projects/medbioinf/data/share/globus/1000g-ont/GRCh38/{sample}/alignments/{sample}.cram',
        ref='/gpfs/project/projects/medbioinf/users/spani/files/ref/GCA_000001405.15_GRCh38_no_alt_analysis_set_maskedGRC_exclusions_v2.fasta'
    output:
        bam=temp('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/GRCh38_BAM/{sample}.bam'),
        ind=temp('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/GRCh38_BAM/{sample}.bam.bai')
    wildcard_constraints:
        sample = "|".join(all_samples)
    conda:
        'envs/preprocessing.yaml'
    resources:
        runtime_hrs=lambda wildcards, attempt: 10 * attempt,
        runtime_min=0,
        mem_total_mb=lambda wildcards, attempt: 5000 * attempt
    shell:
        '''
        seq_cache_populate.pl -root /gpfs/project/projects/medbioinf/users/spani/files/ref /gpfs/project/projects/medbioinf/users/spani/files/ref/GCA_000001405.15_GRCh38_no_alt_analysis_set_maskedGRC_exclusions_v2.fasta
        export REF_PATH=/gpfs/project/projects/medbioinf/users/spani/files/ref/%2s/%2s/%s:http://www.ebi.ac.uk/ena/cram/md5/%s
        export REF_CACHE=/gpfs/project/projects/medbioinf/users/spani/files/ref/%2s/%2s/%s
        samtools view -b -T {input.ref} -o {output.bam} {input.reads}
        samtools index -b {output.bam}
        '''

# haplotagging females separately (since they dont have PAR regions)
rule haplotag_females:
    input:
        reads='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/GRCh38_BAM/{sample}.bam',
        read_ind='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/GRCh38_BAM/{sample}.bam.bai',
        vcf='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}.vcf.gz',
        vcf_ind='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}.vcf.gz.csi',
        ref='/gpfs/project/projects/medbioinf/users/spani/files/ref/GCA_000001405.15_GRCh38_no_alt_analysis_set_maskedGRC_exclusions_v2.fasta'
    output:
        haplotag_list='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/{sample}/{sample}.tsv'
    log:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/{sample}/output.log'
    wildcard_constraints:
        sample = "|".join(female_samples)
    conda:
        'envs/whatshap.yaml'
    resources:
        runtime_hrs=lambda wildcards, attempt: 24 * attempt,
        runtime_min=0,
        mem_total_mb=lambda wildcards, attempt: 80000 * attempt
    shell:
        '''
        seq_cache_populate.pl -root /gpfs/project/projects/medbioinf/users/spani/files/ref /gpfs/project/projects/medbioinf/users/spani/files/ref/GCA_000001405.15_GRCh38_no_alt_analysis_set_maskedGRC_exclusions_v2.fasta
        export REF_PATH=/gpfs/project/projects/medbioinf/users/spani/files/ref/%2s/%2s/%s:http://www.ebi.ac.uk/ena/cram/md5/%s
        export REF_CACHE=/gpfs/project/projects/medbioinf/users/spani/files/ref/%2s/%2s/%s
        whatshap haplotag --reference {input.ref} --sample {wildcards.sample}  --output-haplotag-list {output.haplotag_list} --output /dev/null {input.vcf} {input.reads}
        '''

# Initial haplotagging of male samples (excluding the non-PAR regions)
rule haplotag_males:
    input:
        reads='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/GRCh38_BAM/{sample}.bam',
        read_ind='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/GRCh38_BAM/{sample}.bam.bai',
        vcf='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}_PAR_only.vcf.gz',
        vcf_ind='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/{sample}_PAR_only.vcf.gz.csi',
        ref='/gpfs/project/projects/medbioinf/users/spani/files/ref/GCA_000001405.15_GRCh38_no_alt_analysis_set_maskedGRC_exclusions_v2.fasta'
    output:
        haplotag_list='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/GRCh38/males/{sample}_read_tags.tsv'
    log:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/{sample}/output.log'
    wildcard_constraints:
        sample = "|".join(male_samples)
    conda:
        'envs/whatshap.yaml'
    resources:
        runtime_hrs=lambda wildcards, attempt: 24 * attempt,
        runtime_min=0,
        mem_total_mb=lambda wildcards, attempt: 80000 * attempt
    shell:
        '''
        seq_cache_populate.pl -root /gpfs/project/projects/medbioinf/users/spani/files/ref /gpfs/project/projects/medbioinf/users/spani/files/ref/GCA_000001405.15_GRCh38_no_alt_analysis_set_maskedGRC_exclusions_v2.fasta
        export REF_PATH=/gpfs/project/projects/medbioinf/users/spani/files/ref/%2s/%2s/%s:http://www.ebi.ac.uk/ena/cram/md5/%s
        export REF_CACHE=/gpfs/project/projects/medbioinf/users/spani/files/ref/%2s/%2s/%s
        whatshap haplotag --reference {input.ref} --sample {wildcards.sample} --output-haplotag-list {output.haplotag_list} --output /dev/null {input.vcf} {input.reads}
        '''

# adding the reads mapping to nonPAR regions of the X.
rule add_nonPAR_haplotag_males:
    input:
        haplotag_list='/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/temp/GRCh38/males/{sample}_read_tags.tsv',
        reads='/gpfs/project/projects/medbioinf/data/share/globus/1000g-ont/GRCh38/{sample}/alignments/{sample}.cram',
        script='/gpfs/project/projects/medbioinf/users/spani/scripts/1000g-ont/ont-1kg/snakemake_pipelines/haplotagging/scripts/add_nonPAR_haplotag.py'
    output:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/{sample}/{sample}.tsv'
    wildcard_constraints:
        sample = "|".join(male_samples)
    conda:
        'envs/basic.yaml'        
    resources:
        runtime_hrs=2,
        runtime_min=0,
        mem_total_mb=lambda wildcards, attempt: 40000 * attempt
    shell:
        '''
        seq_cache_populate.pl -root /gpfs/project/projects/medbioinf/users/spani/files/ref /gpfs/project/projects/medbioinf/users/spani/files/ref/GCA_000001405.15_GRCh38_no_alt_analysis_set_maskedGRC_exclusions_v2.fasta
        export REF_PATH=/gpfs/project/projects/medbioinf/users/spani/files/ref/%2s/%2s/%s:http://www.ebi.ac.uk/ena/cram/md5/%s
        export REF_CACHE=/gpfs/project/projects/medbioinf/users/spani/files/ref/%2s/%2s/%s
        python {input.script} --bam {input.reads} --tag-list {input.haplotag_list} --output {output}
        '''

# creating YAML file which contains the list of male and female samples.
rule haplotag_sex_yaml_file:
    input:
        male = expand(male_samples),
        female = expand(female_samples),
        unspecified = expand(unspecified_samples)
    output:
        '/gpfs/project/projects/medbioinf/users/spani/results/1000GP/haplotagging-results/GRCh38/sample_sex.tsv'
    run:
        f = open(output[0], 'w')
        print('male:', file=f)
        for name in input.male:
            print("\t- ", name, file=f)
        print('female:', file=f)
        for name in input.female:
            print("\t- ", name, file=f)
        print('unspecified:', file=f)
        for name in input.unspecified:
            print("\t- ", name, file=f)
        f.close()

