# ABCA7 (https://doi.org/10.1007/s00401-018-1841-z)

## Notes from paper

- Further inspection indicated a 25 bp tandem repeating pattern. The VNTR comprises 592 bp of the human reference genome (chr19:1049437–1050028 [hg19]), corresponding to 23.7 repetitions of the 25 bp VNTR core repeat motif (Fig. 1).
- We observed high variability in VNTR lengths between individuals. The smallest allele was 298 bp (approx. 12 repeat units), which could be confirmed with Sanger sequencing. The largest observed allele size was 10678 bp (approx. 427 tandem repeats).
- Relative to the ABCA7 VNTR in the human reference genome (chr19:1049437–1050028 [hg19], 592 bp), the longest VNTR length causes an eightfold size increase of intron 18 and a 39% extension of ABCA7 pre-mRNA. Increasing VNTR lengths were associated with risk alleles of rs3764650 (p = 6.7 × 10−8) and rs78117248 (p = 6.3 × 10−4) (Fig. 2a and Fig. S5), confirming our observations based on NGS-based approximation.
- We assessed normal variability of VNTR lengths (median = 2388 bp, interquartile range = 2130 bp) within healthy elderly without an rs3764650 risk allele (n = 133) and observed a maximum VNTR length of 5720 bp (Fig. S6).

- For individuals with two alleles within the normal range (< 5720 bp, “wild-type”), distribution was similar between patients and controls (Fig. 2b). However, individuals carrying an allele > 5720 bp (“expanded”), were mostly patients [OR 4.5 (95% CI 1.3–24.2), p = 0.008, Fig. 2b]. Twenty patients (7.3%) carried an expanded VNTR allele in contrast to three controls (1.7%) (Table S2).

Candidate VNTR from our sites list: chr19	1012105	1014401

Conclusion: We have good information regarding the locus and the ONT data shows a greater spread than the HGSVC3 assemblies. Based on the high polymorphism we see in this site, this is a good example where the higher sample size of this study can be beneficial.

# PLIN4 (https://doi.org/10.1007/s00401-020-02164-4)

- All five perilipins share an amphipathic domain composed of an 11-amino acid (aa) sequence, which is particularly extensive in perilipin-4, the 11-mer being repeated three times to generate a 33-mer, in turn repeated 29 or 31 times.
- Since PLIN4 maps to our linked region, we revisited the genomic and transcriptomic patient data and noticed an unusually high coverage in PLIN4 exon 3 (Supplementary Fig. 4)
- PCR amplification of this exon in patient genomic DNA and muscle RNA revealed the wild-type band, and a second ~ 1000 bp higher band (Fig. 2a) not present in unaffected relatives or in 60 ethnic controls. The 31 × 33-aa amphipathic domain of perilipin-4 is encoded by 31 × 99 repetitive sequences in exon 3 [4]
- We amplified cDNA from patient muscle RNA and obtained Oxford Nanopore long-read sequencing, which confirmed that the higher band is an expansion of the normal 31 × 99-nucleotide sequence to 40 × 99 bases, resulting in 297 (9 × 33) extra amino acids (Supplementary Fig. 5).

Candidate VNTR from our sites list: chr19	4494323	4497243

Conclusion: The study provides good information regarding disease-associated VNTR and the statistics obtained from the ONT data and the HGSVC3 assemblies show concordance. This is an interesting case where both datasets agree very well.

# CACNA1C (https://doi.org/10.1016/j.ajhg.2018.07.011)

- We identified one such region (hg38; chr12:2255791–2256090) within the 100 kb interval associated with BD and SCZ. In the most recent human reference genome (hg38), this region is assembled as a tandem repeat composed of ten 30 bp units. Chimpanzees and other non-human primates have a single instance of a homologous 30 bp sequence at this location (Figure 1A), suggesting that an ancestral 30-mer sequence has expanded in the CACNA1C intron during human evolution.
- Polymerase chain reaction (PCR) amplification and sequencing from six chimpanzees confirmed a single instance of a 30-mer sequence, which exactly matches the chimpanzee reference genome. In contrast, when we performed Southern blots on human DNA (see Supplemental Methods), we found restriction fragment sizes consistent with repeat arrays of 3,000 to 30,000+ bp, with the majority of human repeat arrays showing sizes of approximately 6,000 bp (Figures 1B and 1C). We never observed a band size consistent with the human reference genome (hg38). The smallest band size seen in the 181 human samples we assayed (362 alleles) was 10 times larger than the repeat size annotated in the reference assembly (300 bp), while the largest was more than 100 times larger.

- To test whether the length or sequence of the tandem repeat region shows any association with genomic risk markers for BD and SCZ, we examined whole-genome sequence reads from individuals in the 1000 Genomes Project. We limited our analysis to individuals of European or East Asian descent, the two groups in which BD and SCZ risk status has been previously associated with four SNPs clustered in the third intron of the gene (rs2007044, rs1006737, rs4765905, and rs4765913; Figure 2A)
- We first tested whether repeat length is consistently associated with genotype status at the four GWAS SNPs. None of the four SNPs show a significant association with repeat length and the direction of effect is not consistent (Figure S4). It does not appear that repeat array length is associated with the protective or risk genotypes at the GWAS SNPs, at least not in a simple manner.
- We then tested whether specific sequence variants of the 30-mer unit are associated with the risk or protective alleles at the four GWAS SNPs. For each sequence variant of the 30-mer unit, we tested whether its propensity to appear in reads from this repeat region differs between individuals that are homozygous for risk or protective genotypes at the four SNPs (Figure 2B). We identified a number of 30-mer units that are consistently associated with a genotypic class across all four SNPs (Figure S5). When considering only individuals that are homozygous protective at all four GWAS SNPs (“protective haplotype”) or homozygous risk at all four GWAS SNPs (“risk haplotype”), five 30-mer variants are associated with the protective haplotype, and five 30-mer variants are associated with the risk haplotype (Figure 2C). These particular 30-mer units tend to be located in the variable regions observed when aligning the PacBio-sequenced repeat arrays (marked in Figure 1E). Thus, while there is no straightforward association between the overall length of repeat arrays and the risk or protective haplotype, the abundance of particular 30-mer units is significantly and consistently associated with markers for psychiatric disease risk at CACNA1C.

Candidate VNTR from our sites list: chr12	2251595	2257324

Conclusions: This is a very complex site and does not have straightforward association with the disease. In out analysis, we also do not consider all the motifs of this sites since their occurence is very low.

# WDR7 (https://doi.org/10.1016/j.ajhg.2020.07.004)

- we found that squirrel monkeys, marmosets, green monkeys, baboons, rhesus macaques, and gibbons share a 47 bp region, which corresponds to about two-thirds of the 69 bp repeat unit
- Orangutans, chimpanzees, gorillas, and bonobos all have exactly one copy of the full 69 bp sequence in addition to the 47 bp, and the sequence itself is identical between species
- The human reference genome, by contrast, shows 6 copies of the 69 bp sequence, in addition to the 47 bp.
- Overall, this analysis shows that expansion of the repeat is specific to the human lineage.
- The mean and median copy number of the longest allele per individual was also significantly higher in individuals with ALS than control subjects (p = 0.0003; Mann-Whitney test), suggesting that the repeat may modify ALS susceptibility. Notably, the largest repeat length we observed (86 copies) was detected in an individual who developed sALS at age 72.
- SALS samples from the Quebec cohort and the Answer ALS cohort exhibited significantly higher repeat copy number than the Quebec control subjects, as well as the NIAGADS control subjects
- For comparison, the repeat copy number in sporadic Alzheimer disease (AD) samples available through NIAGADS was no different than the control subjects
- Intriguingly, we determined that exactly 6 of the 69 bp in the repeat unit were variable (Figure 2C).
- WDR7 VNTR copy number distribution in the longest amplified alleles from individuals with sALS and sPD and control subjects obtained from the Coriell Institute. Repeat copy number is significantly higher in cases of ALS (median [IQR] = 17.5 [9–24], mean ± SD = 17.7 ± 10.4) versus control subjects (median [IQR] = 15 [8–21], mean ± SD = 15.2 ± 9.66; p = 0.0003; Mann-Whitney test).

Candidate VNTR from our sites list: chr18	57226379	57227527

Conclusion: The study does not provide robust statistics for disease-associated VNTR and hence we are not considering this for now. But it will be interesting to investigate this site later.