configfile: 'config/config.yaml'

include: 'rules/genotyping.smk'
include: 'rules/ont-alignment.smk'
include: 'rules/prepare-vcf.smk'
include: 'rules/bubble-tagging.smk'
include: 'rules/common-rules.smk'
include: 'rules/get-sample-list.smk'

if config['pilot']:
    samples.sort()
    samples=samples[0:2]

wildcard_constraints:
    sample='|'.join(samples)

rule all:
    input:
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/panel/{callset}_filtered.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/panel/{callset}-giggles_filtered.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/panel/{callset}.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/panel/{callset}-giggles.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/panel/giggles-ready_multiallelic.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/panel/giggles-ready_biallelic.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/panel/panel-full_multiallelic.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/panel/panel-full_biallelic.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/genotypes/{sample}-multiallelic.vcf.gz', callset = [c for c in config['callsets'].keys()], sample=samples),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/genotypes/{sample}-biallelic.vcf.gz', callset = [c for c in config['callsets'].keys()], sample=samples),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/genotypes/multisample-multiallelic.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/genotypes/multisample-biallelic.vcf.gz', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/plots/hwe-af/variant-stats.tsv', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/plots/sv-counts/sv_count.tsv', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/plots/hwe-af/hwe.png', callset = [c for c in config['callsets'].keys()]),
        expand('/gpfs/project/projects/medbioinf/users/spani/results/1000GP/augmented_graph/{callset}/plots/qc/audano-curve.png', callset = [c for c in config['callsets'].keys()])